test/runtests.jl

```diff
@@ -2851,6 +2851,71 @@ end
 end
 end
 
+# ≈ doesn't recurse into tuples. this will also check both type and length implicitly
+tplapprox(a::Tuple, b::Tuple) = all(xy -> xy[1] ≈ xy[2], zip(a, b))
+
+@testset "Gradient and Jacobian Outputs" begin
+    @test tplapprox(Enzyme.unstructured_derivative(Enzyme.Forward, x -> 2*x, 3.0), (2.0,))
+    @test tplapprox(Enzyme.unstructured_derivative(Enzyme.Forward, x -> 2*x, 3.0, Val(2)), (2.0,))
+    @test tplapprox(Enzyme.unstructured_derivative(Enzyme.Forward, x -> [x, 2*x], 3.0), ([1.0,2.0],))
+    @test tplapprox(Enzyme.unstructured_derivative(Enzyme.Forward, x -> [x, 2*x], 3.0, Val(2)), ([1.0,2.0],))
+    @test tplapprox(Enzyme.unstructured_derivative(Enzyme.Forward, x -> sum(abs2, x), [2.0, 3.0]), (4.0, 6.0))
+    @test tplapprox(Enzyme.unstructured_derivative(Enzyme.Forward, x -> sum(abs2, x), [2.0, 3.0], Val(2)), (4.0, 6.0))
+    @test tplapprox(Enzyme.unstructured_derivative(Enzyme.Forward, x -> [-x[2], x[1]], [-2.0, 1.0]), ([0.0,1.0],[-1.0,0.0]))
+    @test tplapprox(Enzyme.unstructured_derivative(Enzyme.Forward, x -> [-x[2], x[1]], [-2.0, 1.0], Val(2)), ([0.0,1.0],[-1.0,0.0]))
+
+    x = 3.0
+
+    @test Enzyme.gradient(Enzyme.Forward, x -> x^2, x) ≈ 6.0
+    @test Enzyme.gradient(Enzyme.Reverse, x -> x^2, x) ≈ 6.0
+
+    @test Enzyme.gradient(Enzyme.Forward, x -> [x, x^2], x) ≈ [1.0, 6.0]
+    @test_broken Enzyme.gradient(Enzyme.Reverse, x -> [x, x^2], x) ≈ [1.0, 6.0]
+
+    #NOTE: the discrepancy here is the same issue with the interface that causes there to be much
+    #special handling of gradients and jacobians.  Presumably these should be made consistent at
+    #some point but it would be a breaking change.
+    @test tplapprox(Enzyme.gradient(Enzyme.Forward, x -> sum(2*x), [1.0,1.0]), (2.0, 2.0))
+    @test Enzyme.gradient(Enzyme.Reverse, x -> sum(2*x), [1.0,1.0]) ≈ [2.0, 2.0]
+
+    @test Enzyme.gradient(Enzyme.Forward, x -> 2*x, [1.0,1.0]) ≈ Float64[2 0; 0 2]
+    @test_broken Enzyme.gradient(Enzyme.Reverse, x -> 2*x, [1.0,1.0]) ≈ Float64[2 0; 0 2]
+
+    x = 1.0
```

Comment:
and can we add struct -> scalar
struct -> vector
struct -> matrix
struct -> struct